File: Models/KeyDetectionCustomKey.vb
Namespace Models
    Public NotInheritable Class KeyDetectionCustomKey
        Private ReadOnly _virtualKey As Integer
        Private ReadOnly _scanCode As Integer
        Private ReadOnly _isExtendedKey As Boolean
        Private ReadOnly _displayName As String

        Public Sub New(virtualKey As Integer, scanCode As Integer, isExtendedKey As Boolean, displayName As String)
            _virtualKey = virtualKey
            _scanCode = scanCode
            _isExtendedKey = isExtendedKey
            _displayName = If(displayName, String.Empty)
        End Sub

        Public ReadOnly Property VirtualKey As Integer
            Get
                Return _virtualKey
            End Get
        End Property

        Public ReadOnly Property ScanCode As Integer
            Get
                Return _scanCode
            End Get
        End Property

        Public ReadOnly Property IsExtendedKey As Boolean
            Get
                Return _isExtendedKey
            End Get
        End Property

        Public ReadOnly Property DisplayName As String
            Get
                Return _displayName
            End Get
        End Property

        Public Function Matches(input As RawKeyboardInput) As Boolean
            If input Is Nothing Then
                Return False
            End If

            Return input.VirtualKey = VirtualKey AndAlso
                   input.ScanCode = ScanCode AndAlso
                   input.IsExtendedKey = IsExtendedKey
        End Function

        Public Shared Function FromInput(input As RawKeyboardInput) As KeyDetectionCustomKey
            If input Is Nothing Then
                Return Nothing
            End If

            Return New KeyDetectionCustomKey(input.VirtualKey,
                                             input.ScanCode,
                                             input.IsExtendedKey,
                                             input.DisplayName)
        End Function
    End Class
End Namespace

File: src/ClickSyncMouseTester/Models/KeyDetectionCustomKey.cs
namespace ClickSyncMouseTester.Models;

public sealed class KeyDetectionCustomKey
{
    private readonly int _virtualKey;

    private readonly int _scanCode;

    private readonly bool _isExtendedKey;

    private readonly string _displayName;

    public int VirtualKey => _virtualKey;

    public int ScanCode => _scanCode;

    public bool IsExtendedKey => _isExtendedKey;

    public string DisplayName => _displayName;

    public KeyDetectionCustomKey(int virtualKey, int scanCode, bool isExtendedKey, string displayName)
    {
        _virtualKey = virtualKey;
        _scanCode = scanCode;
        _isExtendedKey = isExtendedKey;
        _displayName = displayName ?? string.Empty;
    }

    public bool Matches(RawKeyboardInput input)
    {
        if (input == null)
        {
            return false;
        }
        return input.VirtualKey == VirtualKey && input.ScanCode == ScanCode && input.IsExtendedKey == IsExtendedKey;
    }

    public static KeyDetectionCustomKey FromInput(RawKeyboardInput input)
    {
        if (input == null)
        {
            return null;
        }
        return new KeyDetectionCustomKey(input.VirtualKey, input.ScanCode, input.IsExtendedKey, input.DisplayName);
    }
}





